pkg/yurthub/util/fs/store.go

```diff
@@ -0,0 +1,288 @@
+/*
+Copyright 2022 The OpenYurt Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package fs
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+)
+
+type ListMode string
+
+var (
+	ListModeDirs  ListMode = "dirs"
+	ListModeFiles ListMode = "files"
+)
+
+type FileSystemOperator struct{}
+
+// Read will read the file at path.
+// If the path does not exist, ErrNotExists will be returned.
+// If the path is not a regular file, ErrIsNotFile will be returned.
+func (fs *FileSystemOperator) Read(path string) ([]byte, error) {
+	if !IfExists(path) {
+		return nil, ErrNotExists
+	}
+
+	if ok, err := IsRegularFile(path); !ok || err != nil {
+		if !ok {
+			return nil, ErrIsNotFile
+		}
+		return nil, err
+	}
+
+	return os.ReadFile(path)
+}
+
+// Write will write the content at path.
+// If the path does not exist, return ErrNotExists.
+// If the path is not a regular file, return ErrIsNotFile.
+func (fs *FileSystemOperator) Write(path string, content []byte) error {
+	if !IfExists(path) {
+		return ErrNotExists
+	}
+
+	if ok, err := IsRegularFile(path); !ok || err != nil {
+		if !ok {
+			return ErrIsNotFile
+		}
+		return err
+	}
+
+	f, err := os.OpenFile(path, os.O_WRONLY|os.O_TRUNC|os.O_SYNC, 0600)
+	if err != nil {
+		return err
+	}
+	n, err := f.Write(content)
+	if err == nil && n < len(content) {
+		err = io.ErrShortWrite
+	}
+	err1 := f.Close()
+	if err == nil {
+		err = err1
+	}
+	return err
+}
+
+// list will list names of entries under the rootDir(except the root dir). If isRecurisive is set, it will
+// walk the path including all subdirs. The returned file names are absolute paths and
+// in lexicographical order.
+// There're two modes:
+// "dirs" returns names of all directories.
+// "files" returns names of all regular files.
+//
+// If the path does not exist, return ErrNotExists.
+// If the path is not a directory, return ErrIsNotDir.
+func (fs *FileSystemOperator) List(rootDir string, mode ListMode, isRecursive bool) ([]string, error) {
+	if !IfExists(rootDir) {
+		return nil, ErrNotExists
+	}
+	if ok, err := IsDir(rootDir); !ok || err != nil {
+		if !ok {
+			return nil, ErrIsNotDir
+		}
+		return nil, err
+	}
+
+	dirs := []string{}
+	files := []string{}
+	if isRecursive {
+		filepath.Walk(rootDir, func(path string, info os.FileInfo, err error) error {
```

Comment:
please use `filepath.WalkDir()` because it's more efficient.

Reply:
fixed